server-timing/load.php

```diff
@@ -113,3 +116,57 @@ function perflab_wrap_server_timing( $callback, $metric_slug, $access_cap ) {
 		return $result;
 	};
 }
+
+/**
+ * Registers the Server-Timing setting.
+ *
+ * @since n.e.x.t
+ */
+function perflab_register_server_timing_setting() {
+	register_setting(
+		PERFLAB_SERVER_TIMING_SCREEN,
+		PERFLAB_SERVER_TIMING_SETTING,
+		array(
+			'type'              => 'object',
+			'sanitize_callback' => 'perflab_sanitize_server_timing_setting',
+			'default'           => array(),
+		)
+	);
+}
+add_action( 'init', 'perflab_register_server_timing_setting' );
+
+/**
+ * Sanitizes the Server-Timing setting.
+ *
+ * @since n.e.x.t
+ *
+ * @param mixed $value Server-Timing setting value.
+ * @return array Sanitized Server-Timing setting value.
+ */
+function perflab_sanitize_server_timing_setting( $value ) {
+	if ( ! is_array( $value ) ) {
+		return array();
+	}
+
+	/*
+	 * Ensure that every element is an indexed array of hook names.
+	 * Any duplicates across a group of hooks are removed.
+	 */
+	return array_filter(
+		array_map(
+			static function( $hooks ) {
+				if ( ! is_array( $hooks ) ) {
+					$hooks = explode( "\n", $hooks );
+				}
+				return array_values(
+					array_unique(
+						array_filter(
+							array_map( 'sanitize_key', $hooks )
```

Comment:
I don't think that `sanitize_key` will work here because hooks can contain non-key characters, even arbitrary characters since hook names often can include variables. Here are some examples from Human Made repos:

[`do_action( 'mercator.mapping.made_primary', $this );`](https://github.com/humanmade/Mercator/blob/9240e7cce9882a9fe452ced8a5cf7845d60c5a2e/class-mapping.php#L138)
[`do_action( 'hm_parse_request_' . $this->get_regex(), $wp );`](https://github.com/humanmade/hm-rewrite/blob/75e6bf7a81f56c8d64a84ff07e4da334ccb4350f/hm-rewrites.php#L187)
[`do_action( 'amf/loaded' );`](https://github.com/humanmade/asset-manager-framework/blob/58e37f495638a40ce6ddab10b400d702cb4883bc/inc/namespace.php#L70)

Reply:
That's a good point. I'm unsure how far we should go with validation to satisfy all sorts of custom cases, arguably the `sanitize_key()` function does work at least for how core uses hooks. I'm thinking maybe I'll change it to a custom regex that also allows for some other "common" hook characters like "/", "." for example. I don't think we need to cover everything here, validation can always be loosened later if needed. Better to be strict and loosen up later rather than the other way around.

Reply:
I implemented a custom regex in https://github.com/WordPress/performance/pull/784/commits/c5fe384ece17f73a7e62837a2b8db3a290ca8a68. As per my previous comment, I acknowledge this probably doesn't encompass _all_ possible hook names, but given there seem to be no real technical limitations, I'm not even sure supporting _all_ possible hook names is a good idea. I _did_ add support for the special characters from your examples, as I believe they are somewhat common in the plugin ecosystem. We can always add more later if needed.